compiler/rustc_middle/src/query/mod.rs

```diff
@@ -1872,6 +1872,13 @@ rustc_queries! {
         eval_always
         desc { "fetching all foreign CrateNum instances" }
     }
+    // Crates that are loaded non-speculatively (not for diagnostics or doc links).
+    // FIXME: This is currently only used for collecting lang items, but should be used instead of
+    // `crates` in most other cases too.
+    query used_crates(_: ()) -> &'tcx [CrateNum] {
+        eval_always
+        desc { "fetching `CrateNum`s for all crates loaded non-speculatively" }
+    }
```

Comment:
Should we directly change the implementation of `crates`?

Reply:
In some cases the new version should not be used (e.g. in metadata encoding due to https://github.com/rust-lang/rust/pull/121167#issuecomment-1947413815), but I didn't audit in which exactly and switched only one case that is necessary for https://github.com/rust-lang/rust/pull/117772.